src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"fmt"
+	re "regexp"
+	"regexp/syntax"
+
+	fstregexp "github.com/m3db/m3/src/m3ninx/index/segment/fst/regexp"
+)
+
+// CompileRegex compiles the provided regexp into an object that can be used to query the various
+// segment implementations.
+func CompileRegex(r []byte) (CompiledRegex, error) {
+	// NB(prateek): We currently use two segment implementations: map-backed, and fst-backed (Vellum).
+	// Due to peculiarities in the implementation of Vellum, we have to make certain modifications
+	// to all incoming regular expressions to ensure compatibility between them.
+
+	// first, we parse the regular expression into the equivalent regex
+	reString := string(r)
+	reAst, err := parseRegexp(reString)
+	if err != nil {
+		return CompiledRegex{}, err
+	}
+
+	// Issue (a): Vellum does not allow regexps which use characters '^', or '$'.
+	// To address this issue, we strip these characters from appropriate locations in the parsed syntax.Regexp
+	// for Vellum's RE.
+	vellumRe, err := ensureRegexpUnanchored(reAst)
+	if err != nil {
+		return CompiledRegex{}, fmt.Errorf("unable to create FST re: %v", err)
+	}
+
+	// Issue (b): Vellum treats every regular expression as anchored, where as the map-backed segment does not.
+	// To address this issue, we ensure that every incoming regular expression is modified to be anchored
+	// when querying the map-backed segment, and isn't anchored when querying Vellum's RE.
+	simpleRe, err := ensureRegexpAnchored(vellumRe)
+	if err != nil {
+		return CompiledRegex{}, fmt.Errorf("unable to create map re: %v", err)
+	}
+
+	var (
+		compiledRegex = CompiledRegex{}
+	)
+	simpleRE, err := re.Compile(simpleRe.String())
+	if err != nil {
+		return compiledRegex, err
+	}
+	compiledRegex.Simple = simpleRE
```

Comment:
nit: Since the previous early return statements have `return CompiledRegex{}, err...` perhaps change this to:
```go
	simpleRE, err := re.Compile(simpleRe.String())
	if err != nil {
		return CompiledRegex{}, err
	}

	compiledRegex := CompiledRegex{Simple: simpleRE}
```

Reply:
sure thing

Reply:
done